org.mozilla.FirefoxNightly.flashplayer/org.mozilla.FirefoxNightly.flashplayer.appdata.xml

```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="runtime">
+  <id>org.mozilla.FirefoxNightly.flashplayer</id>
```

Comment:
Whatever we name should not be specific to FirefoxNightly, I would like to see the extension submitted to Flathub and it can be used by any NPAPI-compatible browser (as well as the Endless OS Flatpak for Firefox which we're shipping). There's no need to rename / re-download the plugin if you change which Flatpak you use.

In any case, I don't think this name is right - we should take one of two approaches. Either we would name it com.adobe.FlashPlayer.NPAPI and we would make the Mozilla Flatpak specifically look for that one extension - given that Mozilla will only load this one plugin now, this has some sense. Alternatively, we would produce a namespace like org.mozilla.Firefox.Plugins.* and name this org.mozilla.Firefox.Plugins.flashplugin - then we'd use mergedirs to allow multiple providers of files in this path.

Reply:
I'm not sure if I understand - lets say we use the com.adobe.FlashPlayer.NPAPI name, wouldn't we still need to create different manifests for different apps using the plugin, as the runtime of the plugin needs to be set to the application providing the extension? Or am I missing something?

Reply:
There is no such requirement to set the runtime of the extension like this, as far as I'm aware. As extensions cannot be run (they are actually equivalent to runtimes inside Flatpak) there is no real meaning of the "runtime" setting in an extension. We can set it to org.freedesktop.Platform or so.

Reply:
We would like to have the Flash Player implemented as an extension of the Firefox application, not an extension of a whole runtime. The way I understand this is that in order to do this, the application itself needs to create the extension point and then the extension itself needs to use the application as a runtime. That is the point I am trying to make - to allow other applications to use the extension, there would need to be an extension point in the runtime (e.g. org.freedesktop.Platform).

Reply:
Yeah, I was actually mistaken. In order to build the extension by itself, it has to extend something that has it defined as an extension point, so it makes sense to specify a Firefox app (that has the extension point), or to have a stub/pointless app which does nothing other than makes the extension as a side-effect when it's built. I prefer to keep the plugin independent of any one Firefox Flatpak if possible, because we already have multiple of them (and will continue to do so - nightly, devedition, ESR, stable, etc).

However, we have another bigger problem - normally for extensions this "runtime" field is only used at build time for setting up the extension build environment in the way we've been discussing - so the extension must join to an app or a runtime specified here in "runtime" and the build environment is set up with either the runtime + extension, or runtime + app + extension mounted. The problem is that with apply_extra, the logic is different - the implementation inside Flatpak seems to assume extra_data is only for apps - so the thing we're installing (an extension here) is mounted as /app, and the thing which is specified in "runtime" is unconditionally mounted as /usr, and /app/bin/apply_extra is executed. So we probably do something crazy like put the extension in /app and Mozilla in /usr, but we have no libc or anything because the runtime isn't mounted anywhere, which is why it fails like this.

I guess we have two ways:
 - fix flatpak so that it does the same at apply_extra time that it does at build time - mounting the extension "where it should go" inside the app or runtime etc, and accept that we have to define and maintain one version of this extension for every app which needs an NPAPI flash plugin. I am responsible at Endless OS for updating our one Flatpak that contains a Flash plugin for Firefox and I really think this approach will drive us all crazy...
 - try and hack around this somehow - perhaps doing a combination of two things - make a stub app which isn't for any purpose other than building the extension - this will let us define the extension point at /app/lib/mozilla/plugins and build the extension there. put the apply_extra script into /app/lib/mozilla/plugins/bin, and try to override the runtime= setting in the extension metadata that is generated - we might be able to do this with just a finish-args of --runtime=org.freedesktop.Platform//1.6. Then when the apply_extra step is done, the extension will appear as /app with a full runtime as /usr, and all we need to do is leave the plugin.so in /app.

I think it should be possible (maybe with finish-args) to end up with the extension

Reply:
I've opened an [issue](https://github.com/flatpak/flatpak/issues/1975) in Flatpak repo regarding this problem (before your explanation here). The description might be a little vague as I wasn't sure what was it exactly I was experiencing with apply_extra. Feel free to add more detailed information.

Not sure if I understand the second approach you propose. If we create a stub app that creates the extension point for the Flash plugin, the plugin.so file can only be copied to the /app folder of the stub app with no way of copying to the /app folder of the actual application using the extension, e.g. FirefoxNightly. Besides, that still doesn't solve the issue with apply_extra, as the script still runs from an extension.

Reply:
An extension is just a folder full of stuff - it's actually the same as a runtime - it doesn't store or require where it's placed. That is entirely under the control of the app which defines the extension point.

So our "stub app" can put it wherever during build time - /app/whatever - nobody actually needs that app - it would just build the extension as a "side-effect" so we just need it somewhere to include the bin/apply_extra script and maybe a symlink from flashplugin.so to extra/flashplugin.so. The Mozilla flatpaks would put it in /app/lib/mozilla/plugins.

The way that we make the apply_extra work is that we know flatpak always puts the extension in /app, so we can provide an /app/bin/apply_extra script, and we override the runtime so that in /usr (during apply extra) we get a real runtime, and our scripts can run then.